File: src/main/java/co/aplicared/jvm/juego/arbúcies/graphics/Sprite.java
package co.aplicared.jvm.juego.arbúcies.graphics;

public class Sprite {

    /**
     * Tile Sprites
     */

    public static Sprite grassSprite = new Sprite(16, 0, 0, SpriteSheet.terrain);
    public static Sprite voidSprite = new Sprite(16, 0x3b9eb5);

    /**
     * Player 1 Sprites
     */
    // Back
    public static Sprite player0Bck0 = new Sprite(32, 0, 0, SpriteSheet.player);
    public static Sprite player0Bck1 = new Sprite(32, 0, 1, SpriteSheet.player);

    // Right
    public static Sprite player0Rght0 = new Sprite(32, 1, 0, SpriteSheet.player);
    public static Sprite player0Rght1 = new Sprite(32, 1, 1, SpriteSheet.player);

    // Front
    public static Sprite player0Frnt0 = new Sprite(32, 2, 0, SpriteSheet.player);
    public static Sprite player0Frnt1 = new Sprite(32, 2, 1, SpriteSheet.player);

    // Left
    public static Sprite player0Left0 = new Sprite(32, 3, 0, SpriteSheet.player);
    public static Sprite player0Left1 = new Sprite(32, 3, 1, SpriteSheet.player);

    public int size, xCoord, yCoord;
    public SpriteSheet sheet;
    public int[] pixels;

    public Sprite(int size, int xCoord, int yCoord, SpriteSheet sheet) {
        this.size = size;
        pixels = new int[size * size];
        this.xCoord = xCoord * size;
        this.yCoord = yCoord * size;
        this.sheet = sheet;
        load();
    }

    public Sprite(int size, int colour) {
        this.size = size;
        pixels = new int[size * size];
        setColour(colour);
    }

    private void setColour(int colour) {
        for (int i = 0; i < size * size; i++) {
            pixels[i] = colour;
        }
    }

    private void load() {
        for (int y = 0; y < size; y++) {
            for (int x = 0; x < size; x++) {
                pixels[x + y * size] = sheet.pixels[(x + xCoord) + (y + yCoord) * sheet.size];
            }
        }
    }
}

File: src/main/kotlin/co/aplicared/jvm/juego/arbucies/graphics/Sprite.kt
package co.aplicared.jvm.juego.arbucies.graphics

class Sprite(val size: Int) {
    var sheet: SpriteSheet? = null
    var xCoord: Int? = null
    var yCoord: Int? = null
    lateinit var pixels: IntArray

    constructor(size: Int, xCoord: Int, yCoord: Int, sheet: SpriteSheet) : this(size) {
        this.xCoord = xCoord * size
        this.yCoord = yCoord * size
        this.sheet = sheet
        pixels = IntArray(size * size)
        load()
    }

    constructor(size: Int, colour: Int) : this(size) {
        pixels = IntArray(size * size)
        setColour(colour)
    }

    private fun setColour(colour: Int) {
        for (i in 0 until size) {
            pixels[i] = colour
        }
    }

    private fun load() {
        for (y in 0 until size) {
            for (x in 0 until size) {
                pixels[x + y * size] = sheet!!.pixels[(x + xCoord!!) + (y + yCoord!!) * sheet!!.size]
            }
        }
    }

    object Tiles {
        val lightGrassSprite = Sprite(16, 0, 0, SpriteSheet.terrain)
        var darkGrassSprite = Sprite(16, 1, 0, SpriteSheet.terrain)
        var dirtSprite = Sprite(16, 2, 0, SpriteSheet.terrain)
        var stoneSprite = Sprite(16, 3, 0, SpriteSheet.terrain)
        var lightInfernoSprite = Sprite(16, 4, 0, SpriteSheet.terrain)
        var darkInfernoSprite = Sprite(16, 5, 0, SpriteSheet.terrain)
        var brickSprite = Sprite(16, 0, 1, SpriteSheet.terrain)
        var waterSprite = Sprite(16, 1, 1, SpriteSheet.terrain)
        var crackedStoneSprite = Sprite(16, 2, 1, SpriteSheet.terrain)
        var mossyCrackedStoneSprite = Sprite(16, 3, 1, SpriteSheet.terrain)
        var birchWoodSprite = Sprite(16, 0, 2, SpriteSheet.terrain)
        var voidSprite = Sprite(16, 0x000)
    }

    object Player {
        // Back
        var player0Bck0 = Sprite(32, 0, 0, SpriteSheet.player)
        var player0Bck1 = Sprite(32, 0, 1, SpriteSheet.player)

        // Right
        var player0Rght0 = Sprite(32, 1, 0, SpriteSheet.player)
        var player0Rght1 = Sprite(32, 1, 1, SpriteSheet.player)

        // Front
        var player0Frnt0 = Sprite(32, 2, 0, SpriteSheet.player)
        var player0Frnt1 = Sprite(32, 2, 1, SpriteSheet.player)

        // Left
        var player0Left0 = Sprite(32, 3, 0, SpriteSheet.player)
        var player0Left1 = Sprite(32, 3, 1, SpriteSheet.player)
    }

    object Projectile {
        var ballProjectile = Sprite(16, 0, 0, SpriteSheet.projectile)
    }
}